doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+
+    @qjit
+    def optimize(init_weights, data, steps):
+        def loss(x):
+            dy = grad(cost, argnum=0)(x, data)[0]
+            return (cost(x, data), dy)
+
+        opt = jaxopt.GradientDescent(loss, stepsize=0.4, value_and_grad=True)
+        update_step = lambda i, *args: tuple(opt.update(*args))
+
+        params = init_weights
+        state = opt.init_state(params)
+
+        return for_loop(0, steps, 1)(update_step)(params, state)[0]
+
+The optimization now takes 574ms ± 43.1ms to complete when using 200 steps.
+Note that, to compute gradients within a qjit-compiled function,
+the :func:`catalyst.grad` function must be used.
+
+JAX support and restrictions
+----------------------------
+
+Catalyst is utilizes JAX for program capture, which means you are able to
+leverage the many functions accessible in ``jax`` and ``jax.numpy`` to write
+code that supports :func:`@qjit <~.qjit>` and dynamic variables.
+
+Currently, we are aiming to support as many JAX functions as possible, however
+there may be cases where there is missing coverage. Known JAX functionality
+that doesn't work with Catalyst includes:
+
+- ``jax.numpy.polyfit``
+- ``jax.numpy.fft``
+- ``jax.debug``
+- ``jax.numpy.ndarray.at[index]`` when ``index`` corresponds to all array
+  indices.
+
+If you come across any other JAX functions that don't work with Catalyst
+(or don't already have a Catalyst equivalents), please let us know by opening
+a `GitHub issue <https://github.com/PennyLaneAI/catalyst/issues>`__.
+
+While leveraging ``jax.numpy`` makes it easy to port over NumPy-based
+PennyLane workflows to Catalyst, we also inherit `various restrictions
+and 'gotchas' from JAX
+<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
+This includes:
+
+* **Pure functions**: Compilation is primarily designed to only work on pure
+  functions. That is, functions that do not have any side-effects; the
+  output is purely dependent only on function inputs.
+
+* **In-place array updates**: Rather than using in-place array updates, the
+  syntax ``new_array = jax_array.at[index].set(value)`` should be used. For
+  more details, see `jax.numpy.ndarray.at
+  <https://jax.readthedocs.io/en/latest/_autosummary /jax.numpy.ndarray.at.html>`__.
+
+* **Lack of stateful random number generators**: In JAX, random number
+  generators need to be explicitly created within the :func:`@qjit <~.qjit>` function
+  using ``jax.random.PRNGKey(int)``:
+
+  >>> @qjit()
+  ... def f():
+  ...     key = jax.random.PRNGKey(0)
+  ...     a = jax.random.normal(key, shape=(1,))
+  ...     return a
+  >>> f()
+  array([-0.78476578])
+
+* **Dynamic-shaped arrays:** Functions that create or return arrays with
+  dynamic shape --- that is, arrays where their shape is determined by a
+  dynamic variable at runtime -- are currently not supported in JAX nor
+  Catalyst. Typically, workarounds involve rewriting the code to utilize
+  ``jnp.where`` where possible.
```

Comment:
Is there a reference for this? I don't think I would know what to do with this hint.

Reply:
@dime10 I don't have a good example no :( This is a tip that they also suggest in the JAX docs (try re-writing using `jnp.where`, which is in fact a wrapper around `lax.select`).

I've also used this in the past (successfully), but I remember it being non-trivial

Reply:
Hmm I see. In this case we can leave it as is I think 👍